### PR TITLE
upstream: exclude draining hosts from eds

### PR DESCRIPTION
**Pull request by milton0825**
Commit Message: upstream: exclude draining hosts from eds
Additional Description: 
Currently we only exclude hosts set to `DRAINING` state via `healthcheck/fail` from load balancing, panic routing threshold calculation. In this PR, we exclude hosts set to `DRAINING` state via EDS as well.
Risk Level: ?
Testing: unit tests
Docs Changes: yes
Release Notes: no

### CODE REVIEW

File: envoy/upstream/upstream.h [hunk 1]
```diff
@@ -149,7 +149,9 @@ class Host : virtual public HostDescription {
   /* is not meant to be routed to. */                                            \
   m(EXCLUDED_VIA_IMMEDIATE_HC_FAIL, 0x80)                                        \
   /* The host failed active HC due to timeout. */                                \
-  m(ACTIVE_HC_TIMEOUT, 0x100)
+  m(ACTIVE_HC_TIMEOUT, 0x100)                                                    \
+  /* The host is currently marked as draining by EDS */                          \
+  m(DRAINING_EDS_HEALTH, 0x200)
```
Comment (wbpcode) [inline envoy/upstream/upstream.h:154]: Basically, it's OK for me to add more EDS statuses here. But we can give some better naming, like `EDS_STATUS_DRAINING`. Then in the future, we may add some new statuses like `EDS_STATUS_TIMEOUT`, etc.

Comment (wbpcode) [reply envoy/upstream/upstream.h:154]: cc @htuch

**Review by wbpcode — COMMENTED** [review]
Thanks for your contribution. And some high level comment are added first.

File: docs/root/intro/arch_overview/operations/draining.rst [hunk 2]
```diff
@@ -13,6 +13,7 @@ Draining occurs at the following times:
 * The server is being :ref:`hot restarted <arch_overview_hot_restart>`.
 * The server begins the graceful drain sequence via the :ref:`drain_listeners?graceful
   <operations_admin_interface_drain>` admin endpoint.
+* The server begins the graceful drain sequence via :ref:`EDS <arch_overview_dynamic_config_eds>`.
```
[the remark this answers is not in the document]
Comment (suniltheta) [reply docs/root/intro/arch_overview/operations/draining.rst:16]: There is a believe in Envoy about eventual consistent service discovery https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/service_discovery.html#on-eventually-consistent-service-discovery

If we are introducing this option then we might need to update on the above document about not rely on health check and solely rely on EDS update with this flag set.

[the remark this answers is not in the document]
Comment (wbpcode) [reply docs/root/intro/arch_overview/operations/draining.rst:16]: Sorry. I still didn't get what you mean. Could make it more clear why `The server begins the graceful drain sequence via EDS`? How the EDS update could effects the server?

[the remark this answers is not in the document]
Comment (milton0825) [reply docs/root/intro/arch_overview/operations/draining.rst:16]: We could set `health_status` of an endpoint in EDS to `draining` and the downstream services getting the EDS update could start draining that endpoint:
https://github.com/envoyproxy/envoy/blob/main/api/envoy/api/v2/endpoint/endpoint_components.proto#L83
https://github.com/envoyproxy/envoy/blob/main/api/envoy/api/v2/core/health_check.proto#L49

[the remark this answers is not in the document]
Comment (milton0825) [reply docs/root/intro/arch_overview/operations/draining.rst:16]: I think `draining` is currently treated as `unhealthy` so downstream services won't be sending new requests to endpoints in `draining` state.

File: api/envoy/admin/v3/clusters.proto [hunk 3]
```diff
@@ -174,6 +174,10 @@ message HostHealthStatus {
   // The host failed active HC due to timeout.
   bool active_hc_timeout = 8;
 
+  // The host should be excluded from panic, spillover, etc. calculations because it was explicitly
+  // taken out of rotation via EDS draining.
+  bool excluded_via_eds_draining = 9;
```
[the remark this answers is not in the document]
Comment (milton0825) [reply api/envoy/admin/v3/clusters.proto]: removed

File: envoy/upstream/upstream.h (hunk 1 above)
Comment (milton0825) [reply envoy/upstream/upstream.h]: renamed to `EDS_STATUS_DRAINING`.

**Comment by milton0825** [comment]
@wbpcode this is ready for another look

File: docs/root/intro/arch_overview/operations/draining.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (wbpcode) [reply docs/root/intro/arch_overview/operations/draining.rst:16]: I think draining of server here is different with draing of single endpoint.

[the remark this answers is not in the document]
Comment (milton0825) [reply docs/root/intro/arch_overview/operations/draining.rst]: Yeah let me remove this comment.

**Comment by KBaichoo** [comment]
friendly ping @wbpcode 🌷

**Review by wbpcode — COMMENTED** [review]
Thanks. LGTM overall. And because the additional excluding is a behavior change. Please add a runtime guard for it and a change log. Thanks.

(And sorry for the delay :(

Hi, cc @htuch, could you also take a quick look to this when you have free time. This is very core code change and I am afraid I may missed something.

**Review by htuch — COMMENTED** [review]
Looks fine to me!

**Review by htuch — DISMISSED** [review]

**Comment by milton0825** [comment]
Is runtime guard needed to get this PR merged?

**Comment by milton0825** [comment]
@wbpcode this is ready for another look

**Comment by milton0825** [comment]
@htuch @wbpcode anything else i need to do to get this PR merged?

**Comment by htuch** [comment]
It was fine on my end, @wbpcode?

**Comment by wbpcode** [comment]
ooooooo, I am so sorry, I missed all these pings. I will take a look now.

**Review by wbpcode — APPROVED** [review]
LGTM. Thanks. And sorry again for my delay and negligence.

File: source/server/admin/clusters_handler.cc [hunk 4]
```diff
@@ -99,6 +99,9 @@ void setHealthFlag(Upstream::Host::HealthFlag flag, const Upstream::Host& host,
     health_status.set_active_hc_timeout(
         host.healthFlagGet(Upstream::Host::HealthFlag::ACTIVE_HC_TIMEOUT));
     break;
+  case Upstream::Host::HealthFlag::EDS_STATUS_DRAINING:
+    health_status.set_eds_health_status(envoy::config::core::v3::DRAINING);
```
Comment (AdiJohn) [inline source/server/admin/clusters_handler.cc:103]: This is wrong. It should be 
health_status.set_eds_health_status(host.healthFlagGet(Host::HealthFlag::EDS_STATUS_DRAINING));

Comment (htuch) [reply source/server/admin/clusters_handler.cc:103]: @milton0825 @wbpcode

Comment (krajshiva) [reply source/server/admin/clusters_handler.cc:103]: https://github.com/envoyproxy/envoy/pull/30622#discussion_r1518125160 is not correct either as healthFlagGet return a bool and set_eds_health_status expects a health status enum. But it is true that there was a bug introduced with this PR which should get fixed by #32801 .

**Pull request merged**